### PR TITLE
[Core | Spark | Integrations] : Fix kryo serialization failure for FileIO

### PR DESCRIPTION
**Pull request by singhpk234**
### About the change

presently spark queries fails when using S3fileIO & GlueCatalog when being used with KryoSerializer ref https://github.com/apache/iceberg/issues/5414#issuecomment-1204319969. This happens because Immutable map part of S3FileIO properties is not serializable with Kryo and seralizers for it not available in Twitter Chill lib as well (which also spark uses). This PR attemps to fix this by using java collections instead which spark will be able to ser/de as it's serializer is present in [Chill lib](https://github.com/twitter/chill/blob/develop/chill-java/src/main/java/com/twitter/chill/java/UnmodifiableMapSerializer.java). 

Solves https://github.com/apache/iceberg/issues/5414

### Testing Done 

Manual Test 

UT to validate the change fix the issue, without the fix the UT fails with exception mentioned in ticket https://github.com/apache/iceberg/issues/5414#issuecomment-1204100668

### CODE REVIEW

**Review by kbendick — APPROVED** [review]
This LGTM.

You might want to put a small comment in the code above the creation of `tableSpecificCatalogProperties` and mention that `ImmutableMap` isn't used to support Kryo serde as that's the normal style.

Totally up to you though.

File: aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java [hunk 1]
```diff
@@ -210,25 +211,24 @@ void initialize(
...
+              tableIdentifier, awsProperties.glueCatalogSkipNameValidation()));
       // FileIO initialization depends on tableSpecificCatalogProperties, so a new FileIO is
       // initialized each time
       return new GlueTableOperations(
           glue,
           lockManager,
           catalogName,
           awsProperties,
-          initializeFileIO(tableSpecificCatalogProperties),
+          initializeFileIO(Collections.unmodifiableMap(tableSpecificCatalogProperties)),
```
Comment (kbendick) [inline aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:231]: To test this, you could do `newTableOps` on a `GlueCatalog` and then ensure that the resulting `FileIO` can be round trip serialized through Kryo (which there's a helper method for I believe).

That way it won't be refactored back to `ImmutableMap`.

Comment (singhpk234) [reply aws/src/main/java/org/apache/iceberg/aws/glue/GlueCatalog.java:232]: +1, having a UT will certainly help us catching this, if this happens again. 

> To test this, you could do newTableOps on a GlueCatalog and then ensure that the resulting FileIO can be round trip serialized through Kryo (which there's a helper method for I believe).

+1 mostly did the same just wrote a new helper for it (not happy with it though, still thinking other alternatives), as at present Kryo helper depends on spark, in background includes Twitter#chill serializers. newTableOps presently have a protected access, so was facing issues in writing spark UT. 

The new helper takes a dependency in twitter-chill which spark uses and i think flink as well (may be wrong here, ref [link](https://flink.apache.org/news/2020/04/15/flink-serialization-tuning-vol-1.html#kryo)).

All suggestions are welcomed here :) !

**Review by rdblue — CHANGES_REQUESTED** [review]
I don't think this is the right way to fix the problem.

The approach in this PR is to avoid passing `ImmutableMap` into `initializeFileIO`, but that makes S3FileIO vulnerable to this problem any time the caller doesn't know not to pass that type of map. Instead, `S3FileIO` should be responsible for making a copy of the incoming map that can be serialized with Kryo.

File: aws/src/test/java/org/apache/iceberg/aws/glue/TestGlueCatalog.java [hunk 2]
```diff
@@ -585,4 +589,21 @@ public void testValidateIdentifierSkipNameValidation() {
         CATALOG_NAME, WAREHOUSE_PATH, props, glue, LockManagers.defaultLockManager(), null);
     Assert.assertEquals(glueCatalog.isValidIdentifier(TableIdentifier.parse("db-1.a-1")), true);
   }
+
+  @Test
+  public void testS3FileIOSerialization() throws IOException {
```
Comment (rdblue) [inline aws/src/test/java/org/apache/iceberg/aws/glue/TestGlueCatalog.java:594]: This should not be a Glue test. It should be a `S3FileIO` test.

Comment (singhpk234) [reply aws/src/test/java/org/apache/iceberg/aws/glue/TestGlueCatalog.java]: Initially added this as a glue test considering only flow of S3FileIO initialization was problematic and newTableOps were protected ... 

adding handling of the immutable map in S3FileIO itself, definitely makes sense.

removed this test and added one for S3FileIO, thanks !

File: aws/src/test/java/org/apache/iceberg/aws/s3/TestS3FileIO.java [hunk 3]
```diff
@@ -259,6 +260,17 @@ public void testFileIOJsonSerialization() {
     }
   }
 
+  @Test
+  public void testS3FileIOSerialization() throws IOException {
```
Comment (rdblue) [inline aws/src/test/java/org/apache/iceberg/aws/s3/TestS3FileIO.java:264]: Can you add `Kryo` to this test case name? Should we also have one for Java serialization?

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
This looks like the right fix to me. Thanks, @singhpk234! Could you also add tests for the other FileIO implementations?

File: aws/src/test/java/org/apache/iceberg/aws/s3/TestS3FileIO.java (hunk 3 above)
Comment (singhpk234) [reply aws/src/test/java/org/apache/iceberg/aws/s3/TestS3FileIO.java]: updated the test case name, added a ut for java serialization as spark test case.

File: build.gradle [hunk 4]
```diff
@@ -379,6 +379,7 @@ project(':iceberg-aws') {
 
     testImplementation 'software.amazon.awssdk:iam'
     testImplementation 'software.amazon.awssdk:s3control'
+    testImplementation 'com.twitter:chill-java:0.10.0'
```
Comment (nastra) [inline build.gradle:382]: should the version number go into `versions.props`?

**Review by nastra — APPROVED** [review]
LGTM, but I would define the version number for `com.twitter:chill-java:0.10.0` in `versions.props`and have `build.gradle` only use the actual artifact (without the version): `testImplementation 'com.twitter:chill-java`

File: build.gradle (hunk 4 above)
Comment (singhpk234) [reply build.gradle]: make sense, moved version.props to versions.props, thanks !

**Review by amogh-jahagirdar — APPROVED** [review]

File: spark/v3.3/build.gradle [hunk 5]
```diff
@@ -87,6 +87,14 @@ project(":iceberg-spark:iceberg-spark-${sparkMajorVersion}_${scalaVersion}") {
...
     testImplementation project(path: ':iceberg-core', configuration: 'testArtifacts')
     testImplementation project(path: ':iceberg-data', configuration: 'testArtifacts')
+    testImplementation project(':iceberg-aws')
+    testImplementation 'software.amazon.awssdk:s3'
+    testImplementation 'software.amazon.awssdk:dynamodb'
+    testImplementation 'software.amazon.awssdk:kms'
+    testImplementation 'software.amazon.awssdk:glue'
+    testImplementation project(':iceberg-gcp')
+    testImplementation platform('com.google.cloud:libraries-bom')
+    testImplementation 'com.google.cloud:google-cloud-storage'
```
Comment (rdblue) [inline spark/v3.3/build.gradle:97]: Aren't these already pulled in through core?

File: build.gradle [hunk 6]
```diff
@@ -379,6 +379,7 @@ project(':iceberg-aws') {
 
     testImplementation 'software.amazon.awssdk:iam'
     testImplementation 'software.amazon.awssdk:s3control'
+    testImplementation 'com.twitter:chill-java'
```
Comment (rdblue) [inline build.gradle:382]: Why is this needed?

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/TestFileIOSerialization.java [hunk 7]
```diff
@@ -101,6 +104,60 @@ public void testHadoopFileIOJavaSerialization() throws IOException, ClassNotFoun
     Assert.assertEquals("Conf values must be present", "v2", actualConf.get("k2"));
   }
 
+  @Test
+  public void testResolvingFileIOKryoSerialization() throws IOException {
```
Comment (rdblue) [inline spark/v3.3/spark/src/test/java/org/apache/iceberg/TestFileIOSerialization.java:108]: Why are these tests in Spark instead of in the module that defines each FileIO?

Comment (singhpk234) [reply spark/v3.3/spark/src/test/java/org/apache/iceberg/TestFileIOSerialization.java:108]: This was  a pre-exisitng test class `TestFileIOSerialization` checking the ser/de for HadoopFileIO, hence added other fileIO's serialization tests also to the same class. 

> the module that defines each FileIO

I thing what we can do here is create the kryoHelper (which doesn't take dependency on spark) in core and since  integrations (i.e aws etc) take a dependency on it, it would be available to all FileIO.

The only concern I had here was that these test can become outdated when spark uses diff version of twitter-chill or uses diff lib like [margo](https://github.com/magro/kryo-serializers).

File: build.gradle (hunk 6 above)
Comment (singhpk234) [reply build.gradle:382]: This is needed because kryo serializer by default is not self-sufficient to de-serialize unmodifiable map, serializers for un-modifiable map comes from twitter's chill library when spark tries to ser/de. Hence to mimic the same behavior without  making aws module taking a dependency on spark, I added this dependency. (more details -  [here](https://spark.apache.org/docs/latest/tuning.html#data-serialization))

File: spark/v3.3/build.gradle (hunk 5 above)
Comment (singhpk234) [reply spark/v3.3/build.gradle:97]: Apologies I didn't follow.

As per my understanding integration (i.e aws / aliyun) takes a dependency on core and the iceberg-spark module doesn't take dependency of them ... later the iceberg-spark-runtime module takes dependency takes `implementation` dependency only on iceberg-aws / iceberg-aliyun ... Iceberg runtime module only has Smoke test.

I tried without these dependencies these tests failed for me. Am I missing something here ?

Comment (singhpk234) [reply spark/v3.3/build.gradle:97]: I think i got what you meant here, gcp module has `implementation` dependency on the above so it would already be available to us .. unlike in aws modules where they are compileOnly ... 

made the changes, thanks !

File: build.gradle (hunk 6 above)
Comment (rdblue) [reply build.gradle:382]: That means that this is not actually serializable with Kryo, then.

Looks like the problem is that you're using `SerializableMap.copyOf(properties).immutableMap()`. Instead, you should store the copy and call `immutableMap()` in the `properties()` method to return the map to the caller.

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/TestFileIOSerialization.java (hunk 7 above)
Comment (rdblue) [reply spark/v3.3/spark/src/test/java/org/apache/iceberg/TestFileIOSerialization.java:108]: Kryo should work out of the box without chill (see my comment above), so it shouldn't be an issue.

Can you add the Kryo test helper to `org.apache.iceberg.TestHelpers`?

File: build.gradle [hunk 8]
```diff
@@ -229,6 +229,7 @@ project(':iceberg-api') {
 
   dependencies {
     implementation project(path: ':iceberg-bundled-guava', configuration: 'shadow')
+    implementation "com.esotericsoftware:kryo"
```
Comment (rdblue) [inline build.gradle:232]: This should be test only. We do not want to add Kryo as a dependency anywhere.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java [hunk 9]
```diff
@@ -156,7 +157,7 @@ public void deleteFile(String path) {
 
   @Override
   public Map<String, String> properties() {
-    return properties;
+    return ((SerializableMap<String, String>) properties).immutableMap();
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java:160]: Why didn't this just change the type of `properties`? It's private.

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/TestFileIOSerialization.java [hunk 10]
```diff
@@ -101,6 +104,60 @@ public void testHadoopFileIOJavaSerialization() throws IOException, ClassNotFoun
     Assert.assertEquals("Conf values must be present", "v2", actualConf.get("k2"));
   }
 
+  @Test
```
Comment (rdblue) [inline spark/v3.3/spark/src/test/java/org/apache/iceberg/TestFileIOSerialization.java:107]: Why are these tests still here? If the test cases moved into individual modules, then there should be no need to re-test in Spark.

Comment (singhpk234) [reply spark/v3.3/spark/src/test/java/org/apache/iceberg/TestFileIOSerialization.java]: Agree with you, i left it here as I wanted to have a test with spark e2e, but what it's testing is not different from what we have already tested in their respective modules. Have removed them.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @singhpk234!

**Comment by singhpk234** [comment]
Thanks @rdblue @kbendick @nastra @amogh-jahagirdar for your awesome reviews :) !!!

**Comment by akshay-kokate-06** [comment]
@singhpk234 I am still facing this issue using `iceberg-spark-runtime-3.3_2.12:1.2.0` and `software.amazon.awssdk:bundle:2.20.18`

**Comment by nastra** [comment]
@akshay-kokate-06 what issue are you seeing exactly? Could you provide the stack trace please?
